File: android/app/src/main/java/com/example/soundauth/deviceList.java
package com.example.soundauth;

import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.*;

import com.example.soundauth.databinding.FragmentDeviceInfoBinding;

public class deviceList extends Fragment {
    private static final String TAG = "DeviceList";
    private FragmentDeviceInfoBinding binding = null;
    private PreferencesManager p;


    public deviceList() {}





    private void removeDevice(DeviceInfo d) {
        var set = p.getDevices();
        set.remove(d);
        p.saveDevices(set);
        updateUI();
    }

    public void addDevice(DeviceInfo d) {
        var set = p.getDevices();
        set.add(d);
        p.saveDevices(set);
        updateUI();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    private void updateUI() {
        var devices = p.getDevices();
        var ctx = requireContext();
        LinearLayout l = binding.list;
        var p = new ViewGroup.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, WRAP_CONTENT);
        l.removeAllViews();
        if (devices.isEmpty()) {
            var tv = new TextView(ctx);
            tv.setLayoutParams(p);
            tv.setText("No devices");
            l.addView(tv);
            return;
        }
        for(var d : devices) {
            var v = new LinearLayout(ctx);
            v.setOrientation(LinearLayout.HORIZONTAL);
            v.setPadding(0,0,8,8);
            v.setGravity(Gravity.CENTER_VERTICAL);

            var tv = new TextView(ctx);
            tv.setText(d.name);
            tv.setLayoutParams(p);

            var btn = new Button(ctx);
            btn.setText("forget");
            btn.setOnClickListener((e)->{
                removeDevice(d);
            });

            tv.setLayoutParams(new LinearLayout.LayoutParams(0, WRAP_CONTENT, 1));
            btn.setLayoutParams(new LinearLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT));


            v.addView(tv);
            v.addView(btn);


            l.addView(v);
        }
        Log.d(TAG, "updateUI: updated with " + l.getChildCount() + " devices.");
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentDeviceInfoBinding.inflate(inflater, container, false);
        SharedPreferences x = requireActivity().getSharedPreferences("prefs", Context.MODE_PRIVATE);
        p = new PreferencesManager(x);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        updateUI();
        LocalBroadcastManager.getInstance(requireContext()).registerReceiver(addBroadcastReceiver, new IntentFilter("device_add"));
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
    }

    @Override
    public void onDetach() {
        LocalBroadcastManager.getInstance(requireContext()).unregisterReceiver(addBroadcastReceiver);
        super.onDetach();
    }

    private final BroadcastReceiver addBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            var data = intent.getByteArrayExtra("device");
            var address = intent.getByteArrayExtra("id");
            assert data != null;
            assert address != null;
            var dev = new DeviceInfo(data, address);
            Log.d(TAG, "newDevice: " + dev);

            addDevice(dev);
        }
    };
}

File: android/app/src/main/java/com/example/soundauth/ui/DeviceList.kt
package com.example.soundauth.ui

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.example.soundauth.DeviceInfo
import com.example.soundauth.PreferencesManager


@Composable
fun DeviceListScreen(
    prefs: SharedPreferences,
    modifier: Modifier = Modifier
) {
    val p = remember { PreferencesManager(prefs) }
    var devices by remember { mutableStateOf(p.getDevices().toList()) }

    fun addDevice(d: DeviceInfo) {
        val set = p.getDevices()
        set.add(d)
        p.saveDevices(set)
        devices = set.toList()
    }

    fun removeDevice(d: DeviceInfo) {
        val set = p.getDevices()
        set.remove(d)
        p.saveDevices(set)
        devices = set.toList()
    }

    // Listen for broadcast receiver (device_add)
    val context = LocalContext.current

    DisposableEffect(Unit) {
        val receiver = object : BroadcastReceiver() {
            override fun onReceive(ctx: Context, intent: Intent) {
                val data = intent.getByteArrayExtra("device")!!
                val secret = intent.getByteArrayExtra("secret")
                val addr = intent.getByteArrayExtra("id")!!

                val dev = DeviceInfo(data, addr).apply {
                    this.secret = secret
                }
                addDevice(dev)
            }
        }

        val filter = IntentFilter("device_add")
        LocalBroadcastManager.getInstance(context).registerReceiver(receiver, filter)

        onDispose {
            LocalBroadcastManager.getInstance(context).unregisterReceiver(receiver)
        }
    }

    // UI
    Column(modifier.padding(12.dp)) {
        if (devices.isEmpty()) {
            Text("No devices")
        } else {
            LazyColumn {
                items(devices) { device ->
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .padding(vertical = 6.dp),
                        horizontalArrangement = Arrangement.SpaceBetween
                    ) {
                        Text(device.name)
                        Button(onClick = { removeDevice(device) }) {
                            Text("Forget")
                        }
                    }
                }
            }
        }
    }
}
